storage/storage_image.go

```diff
@@ -0,0 +1,554 @@
+	keys := []string{}
+	for k, v := range s.BlobData {
+		if err := s.store.SetImageBigData(img.ID, k, v); err != nil {
+			return err
+		}
+		keys = append(keys, k)
+	}
+	for _, key := range keys {
+		delete(s.BlobData, key)
+	}
+	s.ID = img.ID
+	return nil
+}
+
+func (s *storageImage) PutManifest(manifest []byte) error {
+	if err := s.Commit(); err != nil {
+		return err
+	}
+	defer s.saveMetadata()
+	return s.store.SetImageBigData(s.ID, "manifest", manifest)
+}
+
+func (s *storageImage) PutSignatures(signatures [][]byte) error {
+	if err := s.Commit(); err != nil {
+		return err
+	}
+	sizes := []int{}
+	sigblob := []byte{}
+	for _, sig := range signatures {
+		sizes = append(sizes, len(sig))
+		newblob := make([]byte, len(sigblob)+len(sig))
+		copy(newblob, sigblob)
+		copy(newblob[len(sigblob):], sig)
+		sigblob = newblob
+	}
+	s.SignatureSizes = sizes
+	defer s.saveMetadata()
+	return s.store.SetImageBigData(s.ID, "signatures", sigblob)
+}
+
+func (s *storageImage) SupportedManifestMIMETypes() []string {
+	return nil
+}
+
+func (s *storageImage) GetBlob(digest string) (rc io.ReadCloser, n int64, err error) {
+	if blob, ok := s.BlobData[digest]; ok {
+		r := bytes.NewReader(blob)
+		return ioutil.NopCloser(r), r.Size(), nil
+	}
+	if _, ok := s.Layers[digest]; !ok {
+		b, err := s.store.GetImageBigData(s.ID, digest)
+		if err != nil {
+			return nil, -1, err
+		}
+		r := bytes.NewReader(b)
+		logrus.Debugf("exporting opaque data as blob %q", digest)
+		return ioutil.NopCloser(r), r.Size(), nil
+	}
+	logrus.Debugf("exporting filesystem layer as blob %q", digest)
+	return s.diffLayer(s.Layers[digest], true)
+}
+
+func (s *storageImage) diffLayer(layerID string, computeSize bool) (rc io.ReadCloser, n int64, err error) {
+	layer, err := s.store.GetLayer(layerID)
+	if err != nil {
+		return nil, -1, err
+	}
+	layerMeta := storageLayerMetadata{
+		ExpectedSize: -1,
+	}
+	if layer.Metadata != "" {
+		if err := json.Unmarshal([]byte(layer.Metadata), &layerMeta); err != nil {
+			logrus.Errorf("error decoding metadata for layer %q: %v", layerID, err)
+			return nil, -1, err
+		}
+	}
+	if computeSize {
+		if layerMeta.ExpectedSize == -1 {
+			n, err = s.store.DiffSize("", layer.ID)
+			if err != nil {
+				return nil, -1, err
+			}
+		} else {
+			n = layerMeta.ExpectedSize
+		}
+	} else {
+		n = -1
+	}
+	diff, err := s.store.Diff("", layer.ID)
+	if err != nil {
+		return nil, -1, err
+	}
+	return diff, n, nil
+}
+
+func (s *storageImage) GetManifest() (manifest []byte, MIMEType string, err error) {
+	manifest, err = s.store.GetImageBigData(s.ID, "manifest")
+	return manifest, "", err
+}
+
+func (s *storageImage) GetTargetManifest(digest string) (manifest []byte, MIMEType string, err error) {
+	requested, err := ddigest.ParseDigest(digest)
+	if err != nil {
+		return nil, "", err
+	}
+	manifest, err = s.store.GetImageBigData(s.ID, "manifest")
+	manifestDigest := requested.Algorithm().FromBytes(manifest)
+	if manifestDigest != requested {
```

Comment:
AFAICT this can just return `nil, errors.New("Manifest lists are not supported by storage:")`.

`GetTargetManifest` exists to support manifest lists (multi-arch images); if the system only supports a single manifest, this would always fail (because the original `GetManifest` returns the manifest list, and then `GetTargetManifest` is called for an arch-specific manifest from that list, i.e. not the same one as returned by `GetManifest`.

Supporting the ability to get the primary manifest by digest here is not _incorrect_ but it does not help.